lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php

```diff
@@ -43,51 +47,10 @@ class CompositeExpression implements Countable
      */
     public function __construct(string $type, array $parts = [])
```

Comment:
In a future PR, I'd like the constructor to be made `private` in favor of two named constructors — `and()` and `or()` — which would accept `($part, ...$parts)` and therefore not allow construction of an empty composite expression.

Reply:
Yes, I've mentioned that this is temporary in https://github.com/doctrine/dbal/pull/3850#discussion_r370997009. I'd like to complete immutability first, then move on to stop accepting empty parts. And you're right, factory methods will be a good idea!